File: basic_python/practices/Algorithms/04. Dynamic Programming/Number_Substrings_of_1/num_sub.py
class numSub:
    def brute_force(self, s: str) -> int:
        cnt = 0
        for idx in range(len(s)):
            for width in range(1, len(s) + 1 ):
                if s[idx : idx + width] == '1' * width:
                    # print(idx, idx + width, s[idx : idx + width])
                    cnt += 1
        return cnt
    
    def math_trick(self, s: str) -> int:
        """
            Noting that every consecutive 1's form a block, and from a block of length k, 
            the number of valid substrings is: 
                                                k * (k + 1) / 2
        
            Return 
                        sum(accumulate(map(int, s), lambda q,v : -~ q * v))
        """
        ans = 0
        count = 0
        for i in range(len(s)):
            if(s[i] == "1"):
                count += 1
            else:
                ans += ((count * (count + 1)) // 2)     # in case exceed 32-bit % int(10**9 + 7)
                count = 0
        ans += ((count * (count + 1)) // 2)             # % int(10**9 + 7)

        return ans

    def dp(self, s: str) -> int:
        dp = [0] * (len(s) + 1)
        for idx in range(len(s)):
            if s[idx] == "1":
                dp[idx + 1] = dp[idx] + 1
            else:
                dp[idx + 1] = 0
        return sum(dp)

File: basic_python/practices/Algorithms/04. Dynamic Programming/Number_Substrings_of_1/num_sub.cpp
#include <string>
#include <vector>
using namespace std;

class NumSub {
public:

    // ----------------------------------------------------------
    // 1. Brute Force
    // Time: O(n^3) (worst-case)
    // ----------------------------------------------------------
    long long brute_force(const string& s) {
        long long cnt = 0;
        int n = s.size();

        for (int i = 0; i < n; i++) {
            for (int width = 1; width <= n; width++) {
                if (i + width > n) break;               // avoid overflow

                // Check if substring is all '1'
                bool all_one = true;
                for (int k = 0; k < width; k++) {
                    if (s[i + k] != '1') {
                        all_one = false;
                        break;
                    }
                }
                if (all_one) cnt++;
            }
        }
        return cnt;
    }


    // ----------------------------------------------------------
    // 2. Math Trick
    // Count consecutive blocks of 1s
    // For block of length k → contributes k*(k+1)/2 substrings
    // Time: O(n)
    // ----------------------------------------------------------
    long long math_trick(const string& s) {
        long long ans = 0, count = 0;

        for (char c : s) {
            if (c == '1') {
                count++;
            } else {
                ans += (count * (count + 1)) / 2;
                count = 0;
            }
        }

        // last block
        ans += (count * (count + 1)) / 2;

        return ans;
    }


    // ----------------------------------------------------------
    // 3. DP Approach
    // dp[i] = number of consecutive 1s ending at i-1
    // Time: O(n) | Space: O(n)
    // ----------------------------------------------------------
    long long dp(const string& s) {
        int n = s.size();
        vector<long long> dp(n + 1, 0);

        for (int i = 0; i < n; i++) {
            if (s[i] == '1') {
                dp[i + 1] = dp[i] + 1;
            } else {
                dp[i + 1] = 0;
            }
        }

        long long total = 0;
        for (long long x : dp) total += x;

        return total;
    }
};
